third_party/nvfuser/csrc/fusion_segmenter.cpp

```diff
@@ -1759,7 +1763,9 @@ class TranslateApplicableWelford {
 
   //! Translate the given welford op into separate
   //! average and standard deviation calculation.
-  void translateSingleWelford(WelfordOp* welford);
+  //! allow reuse of x - average may cause large persistent buffer
+  //! set default to false
```

Comment:
Could you elaborate?

Reply:
will add comments to the code,
 ```
  //! Translate the given welford op into separate
  //! average and standard deviation calculation.
  //! reuse_x_mean_sub will set persistent buffer to computation data type
  //! which is usually float even the input is in half precision.
  //! set default to false to benefit cases with half precision input.
  void translateSingleWelford(WelfordOp* welford, bool reuse_x_mean_sub = false);
```

Reply:
Can you show a simple example?

Reply:
NVFuserTest.FusionMagicSchedulerLayerNormalization_CUDA with shape of 2048 x 10240
if set `reuse_x_mean_sub = true`, generated kernel will use ` float T3[((10 * 1) * 8)];` to store `input` and then `x_mean_sub`.
if set `reuse_x_mean_sub = false`, generated kernel will use `Array<__half, ((10 * 1) * 8), 8> T30;` to store `input` and `x_mean_sub` will be re-calculated.
======== Generated kernel if `reuse_x_mean_sub = true` ==============
```
__global__ void kernel1(Tensor<__half, 2> T0, Tensor<__half, 1> T2, Tensor<__half, 1> T1, int64_t i3, Tensor<float, 2> T26, Tensor<float, 2> T28, Tensor<__half, 2> T20) {
  alignas(16) extern __shared__ char array[];
  void* shared_mem = array;
  NVFUSER_DEFINE_MAGIC_ZERO
  int i216;
  i216 = ((nvfuser_index_t)blockIdx.x) * i3;
  int i73;
  i73 = ceilDiv((ceilDiv(i3, 8)), 10);
  bool b851;
  b851 = ((nvfuser_index_t)threadIdx.x) < i73;
  bool b195;
  b195 = ((nvfuser_index_t)threadIdx.x) == 0;
  double d15;
  d15 = (double)(i3);
  double d16;
  d16 = 1.00000000000000000e+00 * d15;
  double d24;
  d24 = reciprocal(d16);
  double d48;
  d48 = (double)(i3);
  double d49;
  d49 = 1.00000000000000000e+00 * d48;
  float T3[((10 * 1) * 8)];
  #pragma unroll
  for(nvfuser_index_t i159 = 0; i159 < 10; ++i159) {
    int i219;
    i219 = 8 * (((nvfuser_index_t)threadIdx.x) + (i73 * (i159 + nvfuser_zero)));
    int i220;
    i220 = i216 + i219;
    int i237;
    i237 = 8 * i159;
    bool b854;
    b854 = b851 && ((7 + i219) < i3);
    if (b854) {
      Array<__half, 8, 8> T23;
      T23.set(0);
      loadGlobalToLocal<__half, 8, false>(&T23[0],  &T0[i220]);
      #pragma unroll
      for(nvfuser_index_t i161 = 0; i161 < 8; ++i161) {
        T3[(i237 + i161)]
           = __half2float(T23[i161]);
      }
    } else {
      Array<__half, 8, 8> T23;
      if (b851) {
        T23.set(0);
      }
      if (b854) {
        loadGlobalToLocal<__half, 8, false>(&T23[0],  &T0[i220]);
      }
      #pragma unroll
      for(nvfuser_index_t i161 = 0; i161 < 8; ++i161) {
        if (b851) {
          T3[(i237 + i161)]
             = __half2float(T23[i161]);
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  float T21[1];
  T21[0] = 0.00000000000000000e+00;
  float T31[1];
  if (b851) {
    T31[0] = 0.00000000000000000e+00;
  }
  #pragma unroll
  for(nvfuser_index_t i162 = 0; i162 < 10; ++i162) {
    int i300;
    i300 = 8 * i162;
    bool b984;
    b984 = b851 && ((7 + (8 * (((nvfuser_index_t)threadIdx.x) + (i73 * (i162 + nvfuser_zero))))) < i3);
    if (b984) {
      #pragma unroll
      for(nvfuser_index_t i164 = 0; i164 < 8; ++i164) {
        T31[0]
          = T31[0]
          + T3[(i300 + i164)];
      }
    } else {
      #pragma unroll
      for(nvfuser_index_t i164 = 0; i164 < 8; ++i164) {
        if (b984) {
          T31[0]
            = T31[0]
            + T3[(i300 + i164)];
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  warp::warpReduceTIDX<false>(
    T21[0],
    T31[0],
    [](float &a, float b) { a = a + b; },
    threadIdx,
    blockDim,
    static_cast<float*>(shared_mem),
    true,
    float(0.00000000000000000e+00));
  float T6[1];
  T6[0]
    = T21[0]
    / (float) d49;
  float T9[1];
  broadcast::blockBroadcast<true, false, false>(
    T9[0],
    T6[0],
    static_cast<float*>(shared_mem),
    true);
  float T29[1];
  T29[0]
     = T9[0];
  if (b195) {
    T28[((nvfuser_index_t)blockIdx.x)]
       = T29[0];
  }
  // Alias Allocation - register
  auto& T10 = T3;
  #pragma unroll
  for(nvfuser_index_t i166 = 0; i166 < 10; ++i166) {
    int i384;
    i384 = 8 * i166;
    if (b851) {
      #pragma unroll
      for(nvfuser_index_t i168 = 0; i168 < 8; ++i168) {
        int i385;
        i385 = i384 + i168;
        T10[i385]
          = T3[i385]
          - T9[0];
      }
    } else {
      #pragma unroll
      for(nvfuser_index_t i168 = 0; i168 < 8; ++i168) {
        int i429;
        i429 = i384 + i168;
        if (b851) {
          T10[i429]
            = T3[i429]
            - T9[0];
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  float T7[1];
  T7[0] = 0.00000000000000000e+00;
  float T32[1];
  if (b851) {
    T32[0] = 0.00000000000000000e+00;
  }
  #pragma unroll
  for(nvfuser_index_t i169 = 0; i169 < 10; ++i169) {
    int i467;
    i467 = 8 * i169;
    bool b1115;
    b1115 = b851 && ((7 + (8 * (((nvfuser_index_t)threadIdx.x) + (i73 * (i169 + nvfuser_zero))))) < i3);
    if (b1115) {
      #pragma unroll
      for(nvfuser_index_t i171 = 0; i171 < 8; ++i171) {
        int i468;
        i468 = i467 + i171;
        float T22[1];
        T22[0]
          = T10[i468]
          * T10[i468];
        T32[0]
          = T32[0]
          + T22[0];
      }
    } else {
      #pragma unroll
      for(nvfuser_index_t i171 = 0; i171 < 8; ++i171) {
        int i507;
        i507 = i467 + i171;
        float T22[1];
        if (b851) {
          T22[0]
            = T10[i507]
            * T10[i507];
        }
        if (b1115) {
          T32[0]
            = T32[0]
            + T22[0];
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  warp::warpReduceTIDX<false>(
    T7[0],
    T32[0],
    [](float &a, float b) { a = a + b; },
    threadIdx,
    blockDim,
    static_cast<float*>(shared_mem),
    true,
    float(0.00000000000000000e+00));
  float T11[1];
  broadcast::blockBroadcast<true, false, false>(
    T11[0],
    T7[0],
    static_cast<float*>(shared_mem),
    true);
  float T12[1];
  T12[0]
    = T11[0]
    * (float) d24;
  float T13[1];
  T13[0]
    = T12[0]
    + (float) 9.99999974737875164e-06;
  float T14[1];
  T14[0]
     = rsqrtf(T13[0]);
  #pragma unroll
  for(nvfuser_index_t i183 = 0; i183 < 10; ++i183) {
    int i605;
    i605 = 8 * (((nvfuser_index_t)threadIdx.x) + (i73 * (i183 + nvfuser_zero)));
    int i637;
    i637 = 8 * i183;
    int i676;
    i676 = i216 + i605;
    bool b1306;
    b1306 = b851 && ((7 + i605) < i3);
    if (b1306) {
      Array<__half, 8, 8> T24;
      T24.set(0);
      loadGlobalToLocal<__half, 8, false>(&T24[0],  &T2[i605]);
      Array<__half, 8, 8> T25;
      T25.set(0);
      loadGlobalToLocal<__half, 8, false>(&T25[0],  &T1[i605]);
      // Alias Allocation - register
      auto& T30 = T25;
      #pragma unroll
      for(nvfuser_index_t i182 = 0; i182 < 8; ++i182) {
        float T4[1];
        T4[0]
           = __half2float(T25[i182]);
        float T16[1];
        T16[0]
           = T4[0];
        float T5[1];
        T5[0]
           = __half2float(T24[i182]);
        float T18[1];
        T18[0]
           = T5[0];
        float T15[1];
        T15[0]
          = T10[(i637 + i182)]
          * T14[0];
        float T17[1];
        T17[0]
          = T15[0]
          * T16[0];
        float T19[1];
        T19[0]
          = T17[0]
          + T18[0];
        T30[i182]
           = __float2half(T19[0]);
      }
      loadLocalToGlobal<__half, 8, false>( &T20[i676], &T30[0]);
    } else {
      Array<__half, 8, 8> T24;
      if (b851) {
        T24.set(0);
      }
      if (b1306) {
        loadGlobalToLocal<__half, 8, false>(&T24[0],  &T2[i605]);
      }
      Array<__half, 8, 8> T25;
      if (b851) {
        T25.set(0);
      }
      if (b1306) {
        loadGlobalToLocal<__half, 8, false>(&T25[0],  &T1[i605]);
      }
      // Alias Allocation - register
      auto& T30 = T25;
      #pragma unroll
      for(nvfuser_index_t i182 = 0; i182 < 8; ++i182) {
        float T4[1];
        if (b851) {
          T4[0]
             = __half2float(T25[i182]);
        }
        float T16[1];
        if (b851) {
          T16[0]
             = T4[0];
        }
        float T5[1];
        if (b851) {
          T5[0]
             = __half2float(T24[i182]);
        }
        float T18[1];
        if (b851) {
          T18[0]
             = T5[0];
        }
        float T15[1];
        if (b851) {
          T15[0]
            = T10[(i637 + i182)]
            * T14[0];
        }
        float T17[1];
        if (b851) {
          T17[0]
            = T15[0]
            * T16[0];
        }
        float T19[1];
        if (b851) {
          T19[0]
            = T17[0]
            + T18[0];
        }
        if (b851) {
          T30[i182]
             = __float2half(T19[0]);
        }
      }
      if (b1306) {
        loadLocalToGlobal<__half, 8, false>( &T20[i676], &T30[0]);
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  float T27[1];
  T27[0]
     = T14[0];
  if (b195) {
    T26[((nvfuser_index_t)blockIdx.x)]
       = T27[0];
  }
}
}


```

Reply:
Hmm, it seems to indicate there's some issue with the buffer projection. Thanks for the example. I'll take a look into it.

Reply:
Reusing the existing tensors results in causing `T10` to be persistent, which makes sense as its use paths go through the reduction+broadcast path. Not reusing it does not cause it as it effectively recomputes `T10`. When `T10` is determined as persistent, since it's not possible to be projected, the buffer is allocated as a buffer of `float`, so that's why `T3` is cached as a `float` buffer.

(Projection of persistent buffers means that caching an input as its original type rather than its upcast type. When an input is half, we always cast it to float before doing any arithmetic computations. Our analysis first detects an upcast tensor must be persistent, and then the projection analysis determines if it's possible to cache the input tensor in its original type. This only matters when an input type is lower precision than the type it's actually used)

This is a rather complex unexpected interaction with the translation, and I think it deserves its own PR so that we could easily look back. Could you please create a new separate PR that just disables the reuse? Please also add a unit test. I don't know if there's any case we wanted to reuse, so it should be fine to always not reuse it.

Reply:
updated. will create another PR to disable reuse after this one is merged.  issue number is https://github.com/csarofeen/pytorch/issues/2321